src/rawposix/src/sys_calls.rs

```diff
@@ -0,0 +1,1161 @@
+    arg2_cageid: u64,
+    arg3: u64,
+    arg3_cageid: u64,
+    arg4: u64,
+    arg4_cageid: u64,
+    arg5: u64,
+    arg5_cageid: u64,
+    arg6: u64,
+    arg6_cageid: u64,
+) -> i32 {
+    // would sometimes check, sometimes be a no-op depending on the compiler settings
+    if !(sc_unusedarg(arg2, arg2_cageid)
+        && sc_unusedarg(arg3, arg3_cageid)
+        && sc_unusedarg(arg4, arg4_cageid)
+        && sc_unusedarg(arg5, arg5_cageid)
+        && sc_unusedarg(arg6, arg6_cageid))
+    {
+        return syscall_error(Errno::EFAULT, "fork", "Invalid Arguments");
+    }
+
+    // Modify the fdtable manually
+    fdtables::copy_fdtable_for_cage(child_arg_cageid, child_arg).unwrap();
+
+    // Get the self cage
+    let selfcage = get_cage(child_arg_cageid).unwrap();
+
+    let parent_vmmap = selfcage.vmmap.read();
+    let new_vmmap = parent_vmmap.clone();
+
+    let cageobj = Cage {
+        cageid: child_arg,
+        cwd: RwLock::new(selfcage.cwd.read().clone()),
+        parent: child_arg_cageid,
+        gid: AtomicI32::new(selfcage.gid.load(Relaxed)),
+        uid: AtomicI32::new(selfcage.uid.load(Relaxed)),
+        egid: AtomicI32::new(selfcage.egid.load(Relaxed)),
+        euid: AtomicI32::new(selfcage.euid.load(Relaxed)),
+        rev_shm: Mutex::new(Vec::new()),
+        main_threadid: RwLock::new(0),
+        interval_timer: IntervalTimer::new(child_arg),
+        epoch_handler: DashMap::new(),
+        pending_signals: RwLock::new(vec![]),
+        signalhandler: selfcage.signalhandler.clone(),
+        sigset: AtomicU64::new(0),
+        zombies: RwLock::new(vec![]),
+        child_num:  AtomicU64::new(0),
+        vmmap: RwLock::new(new_vmmap),
+    };
+
+    // increment child counter for parent
+    selfcage.child_num.fetch_add(1, SeqCst);
+
+    add_cage(child_arg, cageobj);
+    0
+}
+
+/// Reference to Linux: https://man7.org/linux/man-pages/man3/exit.3.html
+///
+/// The exit function causes normal process(Cage) termination
+/// The termination entails unmapping all memory references
+/// Removing the cage object from the cage table, closing all open files which is removing corresponding fdtable
+pub fn exit_syscall(
+    cageid: u64,
+    status_arg: u64,
+    status_cageid: u64,
+    arg2: u64,
+    arg2_cageid: u64,
+    arg3: u64,
+    arg3_cageid: u64,
+    arg4: u64,
+    arg4_cageid: u64,
+    arg5: u64,
+    arg5_cageid: u64,
+    arg6: u64,
+    arg6_cageid: u64,
+) -> i32 {
+    let status = sc_convert_sysarg_to_i32(status_arg, status_cageid, cageid);
+    // would sometimes check, sometimes be a no-op depending on the compiler settings
+    if !(sc_unusedarg(arg2, arg2_cageid)
+        && sc_unusedarg(arg3, arg3_cageid)
+        && sc_unusedarg(arg4, arg4_cageid)
+        && sc_unusedarg(arg5, arg5_cageid)
+        && sc_unusedarg(arg6, arg6_cageid))
+    {
+        return syscall_error(Errno::EFAULT, "exit", "Invalid Arguments");
+    }
+
+    let _ = fdtables::remove_cage_from_fdtable(cageid);
```

Comment:
does this logically work? We're removing the cage before doing operations on it 2 lines later?